### PR TITLE
Add matrix factorization example back to pymc3

### PR DESCRIPTION
**Pull request by zaxtax**
Reintroducing the probabilistic matrix factorization example originally from @macks22 with the Movielens dataset instead of Jester. This could still use some polish, but the notebook works albeit slowly.

### CODE REVIEW

**Comment by fonnesbeck** [comment]
This is great, and is already pretty polished, actually. Runs in about 30 min on my MBP.

File: docs/source/notebooks/probabilistic_matrix_factorization.ipynb [hunk 1]
```diff
@@ -0,0 +1,1489 @@
+{
```
Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p>If you run this again, can you bump to PyMC3 v3.8?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p>could be just a single bar plot for this, right? the density plot is a little weird</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p>I'd use</p><p> </p><p><code></code><code></p><p>print(f"Users: {num_users}\nMovies:", {num_items}\nSparsity: {sparsity}")</p><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p>I would replace this with a function:</p><p> </p><p><code></code><code></p><p>def make_pmf_model(train, dim, alpha, std):</p><p>&nbsp;&nbsp;&nbsp;&nbsp;bounds = (1, 5)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;data = train.copy()</p><p>&nbsp;&nbsp;&nbsp;&nbsp;n, m = data.shape</p><p> </p><p>&nbsp;&nbsp;&nbsp;&nbsp;# Perform mean value imputation</p><p>&nbsp;&nbsp;&nbsp;&nbsp;nan_mask = np.isnan(data)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;data[nan_mask] = data[~nan_mask].mean()</p><p> </p><p>&nbsp;&nbsp;&nbsp;&nbsp;# Low precision reflects uncertainty; prevents overfitting.</p><p>&nbsp;&nbsp;&nbsp;&nbsp;# Set to the mean variance across users and items.</p><p>&nbsp;&nbsp;&nbsp;&nbsp;alpha_u = 1 / data.var(axis=1).mean()</p><p>&nbsp;&nbsp;&nbsp;&nbsp;alpha_v = 1 / data.var(axis=0).mean()</p><p> </p><p>&nbsp;&nbsp;&nbsp;&nbsp;# Specify the model.</p><p>&nbsp;&nbsp;&nbsp;&nbsp;logging.info('building the PMF model')</p><p>&nbsp;&nbsp;&nbsp;&nbsp;with pm.Model() as pmf:</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;U = pm.MvNormal(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'U', mu=0, tau=alpha_u * np.eye(dim),</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;shape=(n, dim), testval=np.random.randn(n, dim) * std)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;V = pm.MvNormal(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'V', mu=0, tau=alpha_v * np.eye(dim),</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;shape=(m, dim), testval=np.random.randn(m, dim) * std)</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;R = pm.Normal(</p><p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;'R', mu=(U @ V.T)[~nan_mask], tau=alpha, observed=data[~nan_mask])</p><p>    return pmf</p><p></code><code></code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p>I'd replace the <code>self</code> in these with a <code>pmf</code> instance.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p>i'd update this to use just <code>pm.sample(draws=500, tune=500)</code> (or similar!)</p><p> </p><p>This should automatically use 1 chain per physical core and run diagnostics.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p>this <code>self.std</code> is defined as <code>1 / alpha</code> in the constructor, and is maybe only used here</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p>pmf = make_pmf_model(train, dim=10, alpha=2, std=0.05)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (ColCarroll) [inline docs/source/notebooks/probabilistic_matrix_factorization.ipynb:1]: <p> </p><p><code>with pmf:</code></p><p><code>    map_estimate = pm.find_MAP()</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3709/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='37'/>

**Comment by ColCarroll** [comment]
This is great @zaxtax -- thanks for getting this example back up and running! I can regen the docs so that it is up on the examples site, too.

**Pull request merged**